### PR TITLE
[Performance][Model] DeepSeek-V4: reuse fused post residual for aux hidden states

### PR DESCRIPTION
**Pull request by TeloySXH**
## Performance (measured on 4×H20, sm_90)

Same machine / same recipe (TP4+EP, DSpark K=7 probabilistic + adaptive
verification, KV fp8, think=off, temperature=0), before vs after:

| Scenario | decode tok/s (median) | system tok/s (median) |
| --- | ---: | ---: |
| C4 (4 concurrent, 8 req) | 130.26 → 142.10 (**+9.1%**) | 342.66 → 381.24 (**+11.3%**) |
| C8 (8 concurrent, 16 req) | 109.45 → 114.07 (**+4.2%**) | 492.22 → 571.17 (**+16.0%**) |
| S1 (single, 8 req) | 221.60 → 207.73 (−6.3%) | 188.24 → 178.92 (−5.0%) |
| Long S1 (8k prompt, single) | 229.81 → 217.44 (−5.4%) | 160.07 → 155.24 (−3.0%) |

The win comes from removing per-step HBM traffic that scales with T:
two standalone `mhc_post_tilelang` runs with full `[T,4,H]` writes plus one
`[T,4H]` bf16 copy (~256 MiB of traffic at T=4096). Concurrent steps have
large T and tight HBM, so the wall-clock throughput gains the most (C8
system +16.0%); single-stream decode steps are only ~8 tokens wide, so
nothing is left to save and the small extra `mean` kernel noise shows up
(S1/Long S1 within run-to-run variance, 3 runs per scenario, not an SLA).
TTFT is unchanged (these paths are not on the first-token critical path).
Resident memory is unchanged (the buffer allocation is kept; only the copy
is skipped).

## Why

For DSpark, the aux hidden states (draft `main_proj` inputs) are the
mean-pooled post residuals `[T, hc, H] → [T, H]` of layers 40/41. The old
code ran a standalone `mhc_post_tilelang` for every aux layer even though
the next layer's fused post already produces the exact same post residual.
It also kept the full `[T, hc, H]` MTP buffer copy alive although DSpark
consumes aux states (which are `torch.cat` + `main_proj`), never that
buffer.

## What

- `DeepseekV4DecoderLayer.forward` gains an optional `capture_previous_aux`
  flag; when the *current* layer is an aux layer, the `mean(dim=1)` of the
  fused post residual is captured instead of running a standalone
  `mhc_post_tilelang` for the previous layer.
- The last layer keeps its standalone `mhc_post_tilelang` because
  `hc_head` still needs the full `[T, hc, H]` residual (it also feeds the
  final aux entry).
- The `_mtp_hidden_buffer` copy and the runner's MTP-buffer read are
  skipped when aux hidden states are present (dummy run and real propose
  path).

## Tests

- `tests/kernels/test_mhc_kernels.py`: `.mean(dim=1)` of the fused-post
  residual now matches the reference (4 passed: T ∈ {1,4,8,128}).
- Stability on the internal 4×H20 deployment (v0.28 line): HTTP warmup
  failures=0, DSpark FULL CUDA-Graph captured (80 s, 11.34 GiB/rank),
  bench errors 0/11 runs after / 0/11 before.

> Baseline note: the numbers above were measured on the internal
> vLLM-MoE fork (v0.28 line); this branch is a source-level port to
> current main and GPU revalidation there is pending.

## Duplicate check

GitHub PR search (`aux hidden`, `mhc post`, `DeepSeek-V4`) returns no
existing PR addressing this change.

---
AI assistance was used in preparing this change (see Co-authored-by
trailer in the commit).

### CODE REVIEW

File: vllm/models/deepseek_v4/nvidia/model.py [hunk 1]
```diff
@@ -1420,39 +1433,54 @@ def forward(
...
+                previous_aux,
+            ) = layer(
                 hidden_states,
                 positions,
                 input_ids,
                 post_mix,
                 res_mix,
                 residual,
+                capture_previous_aux=idx in self.aux_hidden_state_layers,
             )
```
[the remark this answers is not in the document]
Comment (TeloySXH) [reply vllm/models/deepseek_v4/nvidia/model.py:1455]: The trigger is shifted by design and produces the same values as before: capture_previous_aux=idx in aux_hidden_state_layers captures at the start of layer idx's forward, where the attn-side fused post computes the post residual of layer idx-1 — the same value the old code computed by running mhc_post_tilelang after forwarding layer idx-1. So aux for a one-based layer id L (dspark_target_layer_ids[i]+1) is stored under key L, and the lookup resolves exactly {41: post40, 42: post41, 43: post42} — the same three values, same order as before. The final-layer branch covers the case where end_layer is the last aux layer (its standalone mhc_post feeds both hc_head and the last aux), matching the old final_aux_recon reuse.

File: vllm/models/deepseek_v4/nvidia/model.py [hunk 2]
```diff
@@ -1464,7 +1492,8 @@ def forward(
         if self.use_sequence_parallel:
             hidden_states = sp_all_gather(hidden_states)[:full_num_tokens]
 
-        if self._mtp_hidden_buffer is not None:
+        # Traditional MTP / no-aux only. DSpark consumes aux hidden states.
+        if self._mtp_hidden_buffer is not None and not aux_hidden_states:
             num_tokens = hidden_states.shape[0]
             self._mtp_hidden_buffer[:num_tokens].copy_(hidden_states.flatten(1))
```
[the remark this answers is not in the document]
Comment (TeloySXH) [reply vllm/models/deepseek_v4/nvidia/model.py:1503]: Good catch — fixed in 458e386: the MTP-buffer copy guard now also checks remote_aux, so on the last PP rank with aux relayed from other ranks the copy is skipped too. (For PP1, DSpark's supported configuration — dspark explicitly rejects PP — remote_aux is empty and behavior is unchanged.)

File: tests/kernels/test_mhc_kernels.py [hunk 3]
```diff
@@ -296,6 +296,12 @@ def run_ref():
     )
 
     torch.testing.assert_close(residual, residual_ref, atol=1e-2, rtol=1e-2)
+    torch.testing.assert_close(
+        residual.mean(dim=1),
+        residual_ref.mean(dim=1),
+        atol=1e-2,
+        rtol=1e-2,
+    )
```
[the remark this answers is not in the document]
Comment (TeloySXH) [reply tests/kernels/test_mhc_kernels.py:304]: Good point — added a CPU-only regression test test_deepseek_v4_capture_previous_aux in the latest commit: it stubs mhc_fused_post_pre_tilelang, calls DeepseekV4DecoderLayer.forward with capture_previous_aux=True, and asserts the 5th return value equals mean(dim=1) of the fused-post residual — and stays None without the flag. This pins the extraction point (attn-side fused post) and the aux ordering semantics at the decoder level.

File: vllm/models/deepseek_v4/nvidia/model.py [hunk 4]
```diff
@@ -1464,7 +1492,8 @@ def forward(
         if self.use_sequence_parallel:
             hidden_states = sp_all_gather(hidden_states)[:full_num_tokens]
 
-        if self._mtp_hidden_buffer is not None:
+        # Traditional MTP / no-aux only. DSpark consumes aux hidden states.
+        if self._mtp_hidden_buffer is not None and not aux_hidden_states:
```
[the remark this answers is not in the document]
Comment (TeloySXH) [reply vllm/models/deepseek_v4/nvidia/model.py]: Already fixed in 458e386 (this review ran on the earlier head): the MTP-buffer copy guard now also checks remote_aux — if self._mtp_hidden_buffer is not None and not (remote_aux or aux_hidden_states).

**Comment by TeloySXH** [comment]
## Local revalidation on current main (this branch)

**PR:** #55575 (`fix/dsv4-aux-hidden-mhc` @ `a3687ada5`)  
**Goal:** GPU revalidation after the v0.28 → current-main port (as noted in the PR body).  
**Not a duplicate of the internal MoE-fork numbers.**

### Environment

| item | value |
| --- | --- |
| Host | 4× NVIDIA H20 (sm_90, 96 GiB), driver 535.161.08 + cuda-compat-12-9 |
| Python | 3.12.14 isolated venv (`/dockerdata/vllm-t-env`), not conda base |
| torch | 2.13.0+cu129 |
| vLLM | editable this tree (`0.1.dev20916+gdc02934a0.precompiled`) |
| Model (smoke) | DeepSeek-V4-Flash-0731, TP4+EP, DSpark K=7 |

Caches/JIT stayed on local XFS. No changes on `main`.

### 1. CPU — aux capture contract

```bash
source env_isolated.sh
export CUDA_VISIBLE_DEVICES=""
python -m pytest --noconftest -v --tb=short \
  tests/kernels/test_mhc_kernels.py::test_deepseek_v4_capture_previous_aux \
  tests/kernels/test_mhc_kernels.py::test_deepseek_v4_mhc_broadcast_finalize_sums_hc_streams \
  tests/kernels/test_mhc_kernels.py::test_deepseek_v4_mhc_broadcast_refit_refreshes_in_place
```

**Result: 3 passed.**

`test_deepseek_v4_capture_previous_aux` initially failed: the stub for `mhc_fused_post_pre_tilelang` was missing the 12th positional (`sinkhorn_repeat`) vs current main. After matching the stub to `*args, **kwargs`, the test passed. The production capture path was not at fault. I can push that one-line test stub fix if wanted.

### 2. GPU — fused post/pre + residual mean (this PR’s kernel assert)

```bash
source env_isolated.sh
python -m pytest --noconftest -v --tb=short \
  tests/kernels/test_mhc_kernels.py::test_mhc_fused_post_pre
```

**Result: 7 passed, 1 failed.**

| case | result |
| --- | --- |
| hc=4, H=4096, T=1/4/8/128 | pass |
| hc=4, H=7168, T=1/4/8 | pass |
| hc=4, H=7168, T=128 | fail |

The **new** assert in this PR (`residual.mean(dim=1)` vs reference) **passed on all 8 cases**, including T=128 H=7168 (rerun: same).

The failure is the **pre-existing** last assert `x` vs `layer_input_ref` (fused pre / layer input), not the residual/mean path:

```
Mismatched elements: 1 / 917504 (0.0%)
Greatest absolute difference: 0.01513671875 at index (2, 3658) (atol=1e-2)
```

Deterministic on rerun (same index / same 0.01514). This is TileLang fused-pre tightness at T=128 H=7168, outside the aux-hidden change. Not treating it as a #55575 blocker.

### 3. Serve smoke (DSpark, so aux-hidden is on the path)

```bash
# isolated env, port 8004 only; adaptive verification off (incompatible with cudagraph NONE)
vllm serve /dockerdata/model/DeepSeek-v4-Flash-0731 \
  --host 127.0.0.1 --port 8004 \
  --tensor-parallel-size 4 --enable-expert-parallel \
  --kv-cache-dtype fp8 --block-size 256 --max-model-len 4096 \
  --gpu-memory-utilization 0.90 \
  --tokenizer-mode deepseek_v4 --trust-remote-code \
  --reasoning-parser deepseek_v4 \
  --speculative-config '{"method":"dspark","num_speculative_tokens":7,"draft_sample_method":"probabilistic","attention_backend":"FLASH_ATTN","enable_adaptive_verification":false}' \
  --compilation-config '{"cudagraph_mode":"NONE"}' \
  --kernel-config '{"enable_jit_warmup":false}' \
  --disable-custom-all-reduce
```

Startup ~8 min (weight load + DeepGEMM warmup 1360 shapes). EngineCore `shm_broadcast` 60s warnings during warmup only.

One greedy request after `Application startup complete`:

```bash
curl -sS http://127.0.0.1:8004/v1/chat/completions \
  -H 'Content-Type: application/json' \
  -d '{"model":"/dockerdata/model/DeepSeek-v4-Flash-0731","messages":[{"role":"user","content":"Reply with the single word: ok"}],"max_tokens":16,"temperature":0}'
```

**Result: HTTP 200.** `choices[0].message.content == "ok"`, `finish_reason=stop`, usage `prompt=90 completion=13` (10 reasoning tokens). No `ERROR` / `Traceback` in the serve log. First-token JIT warnings (`mhc_pre_big_fuse_with_norm_tilelang`, DSpark rejection kernels) are warmup-off noise, not a failure.

Stopped with `./close.sh` (port 8004 only).

Perf table in the PR body is from the **v0.28 fork** (S1/C4/C8/Long S1). This tree is correctness + smoke only; I did not rerun that bench.

### Ask

Please add the `ready` label (or `/ci run`) when convenient. I cannot add it (0 merged PRs).

**Comment by TeloySXH** [comment]
Hi @zyongye @mgoin 👋

I'm TeloySXH, this is my first contribution to vLLM, and I'd be glad to be a long-term contributor going forward. I'd appreciate a review from you both when you have time.

**What it does:** Two standalone `mhc_post_tilelang` runs plus a `[T,4H]` bf16 copy were adding per-step HBM traffic proportional to `T`. This PR captures the fused post-residual of the previous layer inside the attention-side fused post when an aux layer is scheduled, and reuses it for the aux hidden states instead of recomputing — with a CPU-only regression test (`test_deepseek_v4_capture_previous_aux`) pinning the extraction point and ordering.

**Results (4×H20, TP4+EP, DSpark K=7):** system tok/s +11.3% (C4) and +16.0% (C8); single-stream cases are within run-to-run variance. Revalidated on current main and posted in the thread.

I've addressed the earlier review feedback: the `_mtp_hidden_buffer` copy is now also skipped when aux states arrive via `remote_aux` (CodeRabbit/Copilot), and the `capture_previous_aux` indexing (idx vs idx+1) is intentional and produces exactly the same values as the old path — happy to explain or add a comment in code if that would help reviewers. I'm open to any changes you'd suggest, including splitting the PR or dropping the flag if you think simpler is better.

Also, since this is a fork PR and I don't have merged PRs in the repo yet, `pre-run-check` blocks CI — if possible, could you add the `ready` label so the checks can run? Thanks for your time! 🙏  🥰

**Comment by TeloySXH** [comment]
Rebased onto current main. #56633 already landed the same “reuse fused post residual for aux” idea on DeepSeek-V4.1; this PR is still the V4 path (deepseek_v4/nvidia/model.py) plus the model_runner.py skip of _mtp_hidden_buffer when aux states are present, which #56633 explicitly left to this change. The only conflict was appending tests at the end of test_mhc_kernels.py: both the new ROCm fused post/pre cases from main and test_deepseek_v4_capture_previous_aux are kept.

**Comment by TeloySXH** [comment]
Rebased onto current main (`27b7757f`).

This is still the DeepSeek-V4 aux-hidden path (`deepseek_v4/nvidia/model.py`): capture the mean of the previous layer's fused post residual instead of a standalone `mhc_post_tilelang`, and skip `_mtp_hidden_buffer` when aux states are present (including `remote_aux`). #56633 remains the V4.1 line.

Conflicts were only from Mega-Gate landing on the decoder/MTP signatures:
- `DeepseekV4DecoderLayer.forward` keeps `mega_gate_metadata` and still takes `capture_previous_aux`
- FFN still gets `mega_gate_metadata`; the 5th return value is unchanged
- MTP keeps its Mega-Gate routing metadata and unpacks the extra return
- `test_deepseek_v4_capture_previous_aux` stub FFN now accepts `mega_gate_metadata`

No change to the capture point, aux-layer ordering, or the model-runner MTP skip.
